src/Main.elm

```diff
@@ -77,3 +666,640 @@ main =
+       and doesn't allow the country code to start with 0.
+       A plus prefix option is added due to plus being widely used by users for country code prefix instead of 00.
+       The error message kindly points the user into direction of adding a plus followed by numbers only.
+
+       This is a very basic input validation according to E.164, validating that the number contains correct country code
+       and number would require a library or extensive code.
+    -}
+    "^\\+[1-9]\\d{1,14}$"
+        |> Regex.fromString
+        |> Maybe.map
+            (\r ->
+                validateStringNotBlank
+                    |> Validation.andThen (format r)
+                    |> withCustomError "Please fill in the phone number starting with + and following with numbers only."
+            )
+        |> Maybe.withDefault ("Incorrect phone number validation. Please contact your system administrator." |> customError |> fail)
+
+
+{-| Validates that the string is not empty or blank. I think this would be a useful function to include in Scrive/elm-form.
+-}
+validateStringNotBlank : Validation String String
+validateStringNotBlank =
+    customValidation Validation.string
+        (\s ->
+            if isBlank s then
+                customError "This field cannot be empty or consist only of spaces." |> Err
+
+            else
+                Ok s
+        )
+
+
+{-| Validates a field that is a string and that can be empty. I think this would be a useful function to include in Scrive/elm-form,
+maybe with a different name (but emptyString and string are already occupied). I think normal "string" validation shouldn't even include
+is-not-empty validation, unless your best practice is to always wrap your String in Maybe if it is possible for it to be empty...
+-}
+validateString : Validation String String
+validateString v =
+    case Field.asString v of
+        Just s ->
+            Ok s
+
+        Nothing ->
+            Err (Error.value InvalidString)
+
+
+validateContactDetails : Validation String ContactDetails
+validateContactDetails =
+    {-
+       If post is selected then the address (consisting of address, zip, city & country) is mandatory to be filled.
+       If email is selected, then the email is mandatory to be filled.
+       If phone is selected then the phone is mandatory to be filled.
+    -}
+    field (contactDetailsFormFieldToName PreferredContactMethodField) validatePreferredContactMethod
+        |> Validation.andThen
+            (\preferredContactMethod ->
+                let
+                    customErrorForPreferredContactMethod contactMethodName =
+                        contactMethodName ++ " is selected as preferred contact method. Please fill in this field." |> withCustomError
+                in
+                case preferredContactMethod of
+                    Post ->
+                        Validation.map8
+                            ContactDetails
```

Comment:
Is it possible to write this validation without using the record constructor?

The problem with record constructors is that if you change the order of the attributes in the type definition you end up with a very silent bug.

Reply:
This is actually a very old problem now, I was mostly curious about your experience with it.
My solution was a very hacky wrapper around elm-format https://github.com/avh4/elm-format/issues/352#issuecomment-913011486

For testing, we use elm-explorations/test and we're working to be able to use https://package.elm-lang.org/packages/avh4/elm-program-test/latest/

Don't worry about rewriting your code, it's not necessary and honestly it wouldn't be respectful of your time.
We just want to see your experience and your thinking. =)

Reply:
The changing-attribute-order problem bit a previous team of mine in the butt a couple of times.
Nothing horrible, but still, I want my peace of mind. ^_^

To be honest, I didn't even know we had a fork of elm-form until you mentioned it and as far as I can tell we're not using it any more. =|

Sorry for the inability to do edits, we started only now with anonymizing the interviews and we're still figuring it out. :pray:

Reply:
No need to fix it, discussing is enough.

I think we need @raadecki 's review too.